aten/src/ATen/native/cuda/SparseBinaryOpIntersectionKernel.cu

```diff
@@ -16,16 +18,127 @@ struct CUDAKernelLauncher {
 };
 
 struct MulOp {
-  static Tensor apply(const Tensor& a, const Tensor& b) {
-    return a.mul(b);
+  template <typename scalar_t>
+  static FUNCAPI scalar_t apply(scalar_t a, scalar_t b) {
+    return a * b;
+  }
+};
+
+template <>
+FUNCAPI bool MulOp::apply(bool a, bool b) {
+  return a && b;
+}
+
+template <int nt, int vt, typename loop_t>
+C10_LAUNCH_BOUNDS_2(nt, vt)
+__global__ void apply_kernel(int n, loop_t loop) {
+  constexpr int nv = nt * vt;
+  int idx = nv * blockIdx.x + threadIdx.x;
+
+  #pragma unroll
+  for (int i = 0; i < vt; ++i) {
+    if (idx < n) {
+      loop(idx);
+      idx += nt;
+    }
+  }
+}
+
+template <int nt, int vt, typename loop_t>
+void launch_kernel(int64_t n, const loop_t& loop) {
+  TORCH_INTERNAL_ASSERT(0 <= n && n <= std::numeric_limits<int32_t>::max());
+  if (!n) {
+    return;
+  }
+
+  const dim3 block(nt);
+  const dim3 grid((n + block.x * vt - 1) / (block.x * vt));
+  const auto stream = at::cuda::getCurrentCUDAStream();
+  apply_kernel<nt, vt, loop_t><<<grid, block, 0, stream>>>(n, loop);
+  C10_CUDA_KERNEL_LAUNCH_CHECK();
+}
+
+template <typename binary_op_t, typename scalar_t, typename index_t>
+void binary_op_intersection_kernel(
+    TensorIterator& iter,
+    int64_t lhs_nnz_stride,
+    int64_t rhs_nnz_stride) {
+  if (!iter.can_use_32bit_indexing()) {
+    for (auto& sub_iter : iter.with_32bit_indexing()) {
+      binary_op_intersection_kernel<binary_op_t, scalar_t, index_t>(
+          sub_iter, lhs_nnz_stride, rhs_nnz_stride);
+    }
+    return;
+  }
+
+  auto* RESTRICT ptr_res_values_bytes = reinterpret_cast<char*>(iter.data_ptr(0));
+  const auto* ptr_lhs_values_bytes = reinterpret_cast<char*>(iter.data_ptr(1));
+  const auto* ptr_lhs_select_idx_bytes = reinterpret_cast<char*>(iter.data_ptr(2));
+  const auto* ptr_rhs_values_bytes = reinterpret_cast<char*>(iter.data_ptr(3));
+  const auto* ptr_rhs_select_idx_bytes = reinterpret_cast<char*>(iter.data_ptr(4));
+
+  auto offset_calc = make_offset_calculator<5>(iter);
+  auto loop = [=] FUNCAPI (int i) {
+    auto offsets = offset_calc.get(i);
+
+    auto* RESTRICT ptr_res_values = reinterpret_cast<scalar_t*>(ptr_res_values_bytes + offsets[0]);
+    const auto* ptr_lhs_values = reinterpret_cast<const scalar_t*>(ptr_lhs_values_bytes + offsets[1]);
+    const auto lhs_nnz_idx = *reinterpret_cast<const index_t*>(ptr_lhs_select_idx_bytes + offsets[2]);
+    const auto* ptr_rhs_values = reinterpret_cast<const scalar_t*>(ptr_rhs_values_bytes + offsets[3]);
+    const auto rhs_nnz_idx = *reinterpret_cast<const index_t*>(ptr_rhs_select_idx_bytes + offsets[4]);
```

Comment:
@amjames , could you please confirm whether skipping restrict for these value pointers is justified?

Reply:
This is a bit of an odd case, you have assigned RESTRICT to the output pointer. The result used for the TensorIterator is always allocated? If the out variant is called the values are copied to the provided tensor after the fact? 

If we already know that the only write address does not overlap then it should be safe for the compiler to load all read only values once, and writes to the output should not invalidate cached values.

So either restrict should be omitted on the result, or if we can be sure of that address' exclusivity then all pointer may safely be restricted.

Reply:
No allocation in TI, everything is not coalesced. I had everything restricted, but thought maybe const-ness might have been sufficient...